Don't trash userconfig on errors

poelzi opened the pull request:
Instead of opening the current config file for write operations,
write the user config file into a new file and rename after success.

This will no longer trash the config file if the file-system is full.

fixes launchpad bug #1881997

poelzi committed:
Instead of opening the current config file for write operations,
write the user config file into a new file and rename after success.

fixes launchpad bug #1881997

Holzhaus commented on src/preferences/configobject.cpp:
```diff
@@ -211,9 +215,19 @@ template <class ValueType> void ConfigObject<ValueType>::save() {
+                qWarning() << "Could not remove old config file:" << oldConfig.errorString();
+                return false;
+            } else {
+                file.rename(m_filename);
```
This can fail, too. Please check the return value and print a warning.

Holzhaus commented on src/preferences/configobject.cpp:
```diff
@@ -211,9 +215,19 @@ template <class ValueType> void ConfigObject<ValueType>::save() {
+            if (!oldConfig.remove()) {
+                qWarning() << "Could not remove old config file:" << oldConfig.errorString();
+                return false;
+            } else {
```
An else case after `return` is unnecessary.

Holzhaus reviewed (changes requested):
Thanks!

poelzi committed:
Print error when rename fails. Simplify code.

uklotzde commented on src/preferences/configobject.cpp:
```diff
@@ -186,15 +186,19 @@ template <class ValueType> void ConfigObject<ValueType>::reopen(const QString& f
     QReadLocker lock(&m_valuesLock); // we only read the m_values here.
-    QFile file(m_filename);
+    QString tmpFilename(m_filename + ".tmp");
+    QFile file(tmpFilename);
```
tmpFile

uklotzde commented on src/preferences/configobject.cpp:
```diff
@@ -186,15 +186,19 @@ template <class ValueType> void ConfigObject<ValueType>::reopen(const QString& f
     if (!file.open(QIODevice::WriteOnly | QIODevice::Text)) {
-        qDebug() << "Could not write file" << m_filename << ", don't worry.";
-        return;
+        qWarning() << "Could not write config file: " << m_filename;
```
Wrong file path, user tmpFile.fileName()

uklotzde commented on src/preferences/configobject.cpp:
```diff
@@ -186,15 +186,19 @@ template <class ValueType> void ConfigObject<ValueType>::reopen(const QString& f
+bool ConfigObject<ValueType>::save() {
     QReadLocker lock(&m_valuesLock); // we only read the m_values here.
-    QFile file(m_filename);
+    QString tmpFilename(m_filename + ".tmp");
```
The file extension should be defined as a constant

uklotzde commented on src/preferences/configobject.cpp:
```diff
@@ -211,9 +215,21 @@ template <class ValueType> void ConfigObject<ValueType>::save() {
+                qWarning() << "Could not remove old config file: " << oldConfig.errorString();
+                return false;
+            }
+            if (file.rename(m_filename)) {
```
Continue with the early return pattern from the preceding conditional here, i.e. `if (!file.rename(m_filename)) {`. Following different branching styles (success vs failure) is difficult otherwise.

poelzi committed:
Also check stream position when writing config

Calculate a minimum size the config must have and check the seek position after
flushing.
Check filesize on disk with expected size.

poelzi commented:
I did some more testing and added additional sanity checks.
It now also properly catches cases of partially written config.
Also print the configfile name as well where the problem occurred.

Holzhaus commented on src/preferences/configobject.cpp:
```diff
@@ -186,34 +187,66 @@ template <class ValueType> void ConfigObject<ValueType>::reopen(const QString& f
+    QTextStream stream(&tmpFile);
+    stream.setCodec("UTF-8");
+
+    QString grp = "";
```
group (i think we can afford the 2 additional characters)

Holzhaus commented on src/preferences/configobject.cpp:
```diff
@@ -13,6 +13,7 @@
 
 // TODO(rryan): Move to a utility file.
 namespace {
+const char* kTempFilenameExtension = "-tmp";
```
```suggestion
const QString kTempFilenameExtension = QStringLiteral(".tmp");
```

Holzhaus commented on src/preferences/configobject.cpp:
```diff
@@ -186,34 +187,66 @@ template <class ValueType> void ConfigObject<ValueType>::reopen(const QString& f
+    tmpFile.close();
+    if (tmpFile.error() !=
+            QFile::NoError) { //could be better... should actually say what the error was..
+        qWarning() << "Error while writing configuration file: "
```
```suggestion
        qWarning().nospace() << "Error while writing configuration file: "
```

Holzhaus commented on src/preferences/configobject.cpp:
```diff
@@ -186,34 +187,66 @@ template <class ValueType> void ConfigObject<ValueType>::reopen(const QString& f
+
+    QFile oldConfig(m_filename);
+    if (!oldConfig.remove()) {
+        qWarning() << "Could not remove old config file: "
```
```suggestion
        qWarning().nospace() << "Could not remove old config file: "
```

Holzhaus commented on src/preferences/configobject.cpp:
```diff
@@ -186,34 +187,66 @@ template <class ValueType> void ConfigObject<ValueType>::reopen(const QString& f
+        return false;
+    }
+    if (!tmpFile.rename(m_filename)) {
+        qWarning() << "Could not rename tmp file to config file: "
```
```suggestion
        qWarning().nospace() << "Could not rename tmp file to config file: "
```

poelzi committed:
style fixes

uklotzde commented on src/preferences/configobject.cpp:
```diff
@@ -186,34 +187,66 @@ template <class ValueType> void ConfigObject<ValueType>::reopen(const QString& f
 
-        QString grp = "";
+    stream.flush();
+    tmpFile.flush();
```
Duplicate flush

> If you call flush(), QTextStream will empty all data from its write buffer into the device and call flush() on the device.

Holzhaus replied:
same here, this isn't resolved yet.

uklotzde replied:
@poelzi I am not going to do any future reviews if comments are ignored and I have to double check everything again. Not possible.

Holzhaus replied:
Maybe he just forgot to push?

uklotzde replied:
Yes, probably ;) I must admit that I marked issues as resolved before pushing.

Holzhaus replied:
Same.

poelzi replied:
yes :(

uklotzde commented on src/preferences/configobject.cpp:
```diff
@@ -186,34 +187,66 @@ template <class ValueType> void ConfigObject<ValueType>::reopen(const QString& f
+
+    QString group = "";
+
+    uint minLength = 0;
```
signed/unsigned mismatch, use `qint64` like pos()

I wonder why you didn't notice any warnings or build failures using -DWARNINGS_FATAL=ON as suggested?

poelzi replied:
since floating convresion got a warning this is not possible anymore. my env did not print any type warning otherwise I would have changed it.

Holzhaus replied:
> since floating convresion got a warning this is not possible anymore.

What do you mean? Are there still float-conversion warnings in the code? Latest 2.3/main should be clean, you can `git pull` the changes into this branch to get all fixes from upstream.

poelzi replied:
I merged it before. Did you disable ccache ? I get at least a dozen warnings when it has to compile many files.

poelzi replied:
https://gist.github.com/poelzi/a1acac5d0a2109c5b6ed890cd422cd18

Holzhaus replied:
Weird, what compiler + version are you using? Yes, I disabled ccache and get no warnings with gcc 10.2.0. Also, our CI builds also don't warn about these conversions even though they have `WARNINGS_FATAL` enabled.

I'll look into fixing these errors.

poelzi replied:
I currently compile with clang version 7.1.0 (tags/RELEASE_710/final)

Holzhaus replied:
I'll try to set up clang, but our CI is using clang as well and doesn't complain.

I've gone through your log and hopefully fixed all warnings: https://github.com/mixxxdj/mixxx/pull/3227

Please test.

uklotzde replied:
Compiling with Clang 10 did not show any new errors besides #3152.

uklotzde replied:
Strange, newer compilers often are more strict than old versions.

Holzhaus replied:
@poelzi Why did you mark this as resolved?

poelzi replied:
because I forgot to push :(

poelzi committed:
change counter type, remove double flush

uklotzde commented on src/preferences/configobject.cpp:
```diff
@@ -186,34 +187,65 @@ template <class ValueType> void ConfigObject<ValueType>::reopen(const QString& f
-        QString grp = "";
+    stream.flush();
+    // the stream is usually longer, depending on the amount of encoded data.
+    if (stream.pos() < minLength || QFileInfo(tmpFile).size() != stream.pos()) {
```
Do we really need the redundant minLength check if the stream already keeps track of the position? Why and how could the stream position ever fall behind minLength??

It adds a lot of redundant code that needs to be maintained and it is unknown if it ever catches any inconsistency.

poelzi replied:
Either take it or leave it now. I will not put any more work in this now. It works, it is safe for all cases I could test. Even partial writes of the file.

poelzi replied:
Without the minLength check you need to check some other way for the safeConfig object to not cause errors being reported when writing the empty config file.

uklotzde replied:
I still don't get the reasoning. Why is checking QTextStream::status() for QTextStream::WriteFailed after flushing not sufficient to detect any write errors?

poelzi replied:
Because there is a safemode instance running that writes a 0 byte file by default and triggers the error handler. Instead of setting a bool from 0 to 1 I decided to do extra safety and count what is at least expected. For me, loosing settings is super annoying, so I usually put them to a safe space. I have decades of experience with computer systems to know that a full file systems behaviour depends on the system implementation.
This is the nit picking be mentioned before, that is also bothering me very much.

uklotzde replied:
@daschuer @Be @Holzhaus Could somone else please comment on this? Does the code reveal all those thoughts? I suppose that any other developer stumbling over this code in the future will ask the same questions I was asking.

We rely on Qt for doing the system abstractions for us. If this doesn't work as expected and we need to introduce workarounds that are basically mistrusting and bypassing the public Qt API, then we should document those exceptional behaviors and the corresponding mitigations clearly.

Holzhaus replied:
I agree, there should at least be a comment why this is necessary in the code. Otherwise the next dev that touches this might just remove it, because it looks redundant.

uklotzde replied:
Thanks for adding a comment!

Recommended reading (off the record): http://danluu.com/file-consistency/

TL;DR: Use SQLite as a good default if you need to store data reliably.

poelzi committed:
Add comment about minLenth counter

uklotzde reviewed (approved):
Haven't tested it yet, but code-wise LGTM so far.

uklotzde commented:
@Holzhaus Ready? No issues so far.

Holzhaus commented:
> @Holzhaus Ready? No issues so far.

Ok for me.

uklotzde commented:
Thank you! LGTM

ronso0 commented:
@poelzi 
there 's now https://bugs.launchpad.net/mixxx/+bug/1903626 about not saving the beatgrid opacity, and I can confirm Mixxx doesn't save any skin config. Didn't proceed testing other configurations but I guess this PR is the cause.
I revert the commits one by one and it's the very first that causes the issue 8eb499b57b9ec22c36fa5fb2c1df290a1511e17b
Please provide a quickfix.

Holzhaus commented:
I also notices these warnings on every shutdown:

```
Warning [Main]: Could not remove old config file: "/home/jan/.mixxx/mixxx.cfg": "No such file or directory"
Warning [LibraryScanner 1]: QSqlDatabasePrivate::removeDatabase: connection 'MIXXX-2' is still in use, all queries will cease to work.
Warning [Main]: Could not remove old config file: "/home/jan/.mixxx/mixxx.cfg": "No such file or directory"
Warning [Main]: Could not remove old config file: "/home/jan/.mixxx/sandbox.cfg": "No such file or directory"
```

ronso0 commented:
_ping_ @poelzi 
We need a fix for this!

ronso0 commented:
@poelzi Please take a look at the error.
Currently it doesn't make sense to point Linux users to current 2.3 builds for troubleshooting if the config is not saved.

Holzhaus commented:
Revert for now?

ronso0 commented:
taking a look at this right now, trying to find out why it can't remove the old config.

ronso0 commented:
Hmm..suddenly it works for me here, all preferences and skin settings are saved.
maybe changes to 2.3 in the meantime fixed it??
@Holzhaus Can you confirm it works for you, too?

uklotzde commented:
Unfortunately, I cannot reproduce the issues on Fedora and Qt 5.15. No idea what is wrong with these changes!?

Be-ing commented:
> I also noticed these warnings on every shutdown:

Fix is here: #3371
